[FX][EZ] Allow constructing GraphModule with dict for root

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44679 [FX][EZ] Allow constructing GraphModule with dict for root**

When generating a GraphModule programmatically (i.e. not from symbolic_trace)(e.g. when splitting up a `GraphModule` into multiple `GraphModules`), it's pretty annoying to materialize a full Module hierarchy before instantiating the `GraphModule` just to copy over attributes into the right place. This change makes it so that `root` on the `GraphModule` constructor can take a dictionary mapping qualified names to the attributes that should be stored at those places in the GraphModule's hierarchy.

The GraphModule in the test case ends up looking like:

```
GraphModuleImpl(
  (foo): Module(
    (bar): Module(
      (baz): Linear(in_features=3, out_features=4, bias=True)
    )
  )
  (zip): Module(
    (zap): Module()
  )
)
def forward(self, x):
    foo_bar_baz = self.foo.bar.baz(x)
    zip_zap_zam = self.zip.zap.zam
    add_1 = foo_bar_baz + zip_zap_zam
    

    return add_1

```


Differential Revision: [D23696766](https://our.internmc.facebook.com/intern/diff/D23696766)

Commit:
[FX][EZ] Allow constructing GraphModule with dict for root

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -82,6 +82,26 @@ def _copy_attr(from_module: torch.nn.Module, to_module: torch.nn.Module, target:
+    *prefix, field = target.split('.')
+    for item in prefix:
+        t = getattr(to_module, item, None)
+        if from_obj is t:
```
I don't think it can ever happen (because `from_obj` would be a thing at different level of hierarchy)

But it brings an important point: you should either ban nested names to appear in the dict at the same time (i.e. `foo` and `foo.bar`) or handle it properly in which case you need to enforce the ordering in which you do assignments below

jamesr66a replied:
@dzhulgakov updated to sort from shortest to longest qualified names. I think this should do the right thing, let me know if there's something screwy

Commit:
Update on "[FX][EZ] Allow constructing GraphModule with dict for root"


When generating a GraphModule programmatically (i.e. not from symbolic_trace)(e.g. when splitting up a `GraphModule` into multiple `GraphModules`), it's pretty annoying to materialize a full Module hierarchy before instantiating the `GraphModule` just to copy over attributes into the right place. This change makes it so that `root` on the `GraphModule` constructor can take a dictionary mapping qualified names to the attributes that should be stored at those places in the GraphModule's hierarchy.

The GraphModule in the test case ends up looking like:

```
GraphModuleImpl(
  (foo): Module(
    (bar): Module(
      (baz): Linear(in_features=3, out_features=4, bias=True)
    )
  )
  (zip): Module(
    (zap): Module()
  )
)
def forward(self, x):
    foo_bar_baz = self.foo.bar.baz(x)
    zip_zap_zam = self.zip.zap.zam
    add_1 = foo_bar_baz + zip_zap_zam
    

    return add_1

```


Differential Revision: [D23696766](https://our.internmc.facebook.com/intern/diff/D23696766)

[ghstack-poisoned]

Commit:
Update on "[FX][EZ] Allow constructing GraphModule with dict for root"


When generating a GraphModule programmatically (i.e. not from symbolic_trace)(e.g. when splitting up a `GraphModule` into multiple `GraphModules`), it's pretty annoying to materialize a full Module hierarchy before instantiating the `GraphModule` just to copy over attributes into the right place. This change makes it so that `root` on the `GraphModule` constructor can take a dictionary mapping qualified names to the attributes that should be stored at those places in the GraphModule's hierarchy.

The GraphModule in the test case ends up looking like:

```
GraphModuleImpl(
  (foo): Module(
    (bar): Module(
      (baz): Linear(in_features=3, out_features=4, bias=True)
    )
  )
  (zip): Module(
    (zap): Module()
  )
)
def forward(self, x):
    foo_bar_baz = self.foo.bar.baz(x)
    zip_zap_zam = self.zip.zap.zam
    add_1 = foo_bar_baz + zip_zap_zam
    

    return add_1

```


Differential Revision: [D23696766](https://our.internmc.facebook.com/intern/diff/D23696766)

[ghstack-poisoned]

zdevito reviewed (approved):
Sounds like a good idea to me.

Commit:
Update on "[FX][EZ] Allow constructing GraphModule with dict for root"


When generating a GraphModule programmatically (i.e. not from symbolic_trace)(e.g. when splitting up a `GraphModule` into multiple `GraphModules`), it's pretty annoying to materialize a full Module hierarchy before instantiating the `GraphModule` just to copy over attributes into the right place. This change makes it so that `root` on the `GraphModule` constructor can take a dictionary mapping qualified names to the attributes that should be stored at those places in the GraphModule's hierarchy.

The GraphModule in the test case ends up looking like:

```
GraphModuleImpl(
  (foo): Module(
    (bar): Module(
      (baz): Linear(in_features=3, out_features=4, bias=True)
    )
  )
  (zip): Module(
    (zap): Module()
  )
)
def forward(self, x):
    foo_bar_baz = self.foo.bar.baz(x)
    zip_zap_zam = self.zip.zap.zam
    add_1 = foo_bar_baz + zip_zap_zam
    

    return add_1

```


Differential Revision: [D23696766](https://our.internmc.facebook.com/intern/diff/D23696766)

[ghstack-poisoned]
